<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Introduction" data-toc-modified-id="Introduction-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introduction</a></span></li><li><span><a href="#BigQuery-Web-UI" data-toc-modified-id="BigQuery-Web-UI-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>BigQuery Web UI</a></span></li><li><span><a href="#BigQuery-API" data-toc-modified-id="BigQuery-API-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>BigQuery API</a></span><ul class="toc-item"><li><span><a href="#Authentication" data-toc-modified-id="Authentication-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Authentication</a></span></li><li><span><a href="#Using-the-Python-API" data-toc-modified-id="Using-the-Python-API-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Using the Python API</a></span><ul class="toc-item"><li><span><a href="#Final-Notes" data-toc-modified-id="Final-Notes-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Final Notes</a></span></li><li><span><a href="#Next-Steps" data-toc-modified-id="Next-Steps-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Next Steps</a></span></li></ul></li></ul></li><li><span><a href="#EDA" data-toc-modified-id="EDA-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>EDA</a></span></li><li><span><a href="#--END--" data-toc-modified-id="--END---5"><span class="toc-item-num">5&nbsp;&nbsp;</span>- END -</a></span></li></ul></div>

![alt text](https://i.imgur.com/1WaY7aA.png)



---



---



# Lab 2.2.4 
# *The Google BigQuery UI and API*

## Introduction

The Google BigQuery UI provides access to Google's extensive collection of public data sets via an SQL-based query engine. 

The BigQuery API provides programmatic access to the data sets.

We can use the UI to discover interesting data before writing Python code to access it. Then we can reproduce it in an API request so as to aggregate large amounts of data on Google's infrastructure before pulling the results into our application.

## BigQuery Web UI

Work through the Quickstart at https://cloud.google.com/bigquery/docs/quickstarts/quickstart-web-ui.

You will need to set up a Google Cloud Platform account if you don't already have one. (This should not cost anything during the trial period unless you perform a large amount of querying. Afterwards, costs are based on actual resource usage, but most offerings have a free tier.)

## BigQuery API

You should already have the Google Cloud Client Library for Python installed (https://cloud.google.com/python/setup).

- Open Google Cloud Console (https://console.cloud.google.com/home/) and select to create a project.
- Under "Getting Started", select "Enable APIs and get credentials such as keys".

- In the API table, make sure the BigQuery API is enabled. Page back.

### Authentication

Go to https://cloud.google.com/docs/authentication/production and click the button to create a service account.


- Fill out the form, giving the account an appropriate name, and choose "Project Owner" for Account Type.

- Click "Create".

- The keys will get saved to your computer:
- Note the location and copy the file path (of the json file) to somewhere safe, for future reference.

- See here for more information:

https://cloud.google.com/iam/docs/understanding-service-accounts?&_ga=2.173177830.-495703703.1532572448#managing_service_account_keys 

This is supposed to get implicit key retrieval working:

- Windows:
    `set GOOGLE_APPLICATION_CREDENTIALS=[PATH]`
    
- Linux, MacOS:
    `export GOOGLE_APPLICATION_CREDENTIALS=[PATH]`
    
where `[PATH]` is the full file path of your json key file.

### Using the Python API

Google provides Python libraries for wrapping the Google APIs. For conda users, these are available on the "conda-forge" channel. (There are other Python libraries for Google APIs, possibly adding novel features or ease of use, but Google's is presumably the most current.)

(Installing the "google-cloud-storage" and "google-cloud-bigquery" libraries should cover all the dependences for this lab.)

In [ ]:
# conda install -c conda-forge google-cloud-storage
# conda install -c conda-forge google-cloud-bigquery

In [1]:
from google.cloud import bigquery
from google.cloud.bigquery import Dataset
from google.cloud import storage

If you have managed to get implicit key retrieval working, you can call `.Client()` with no argument:

In [2]:
storage_client = storage.Client()

If you aren't so lucky, you need to invoke a method of the `.Client` object that takes the path to your key files as a string argument:

In [3]:
#key_path = '[PATH]'  # put the path to your json key file here 
#,                    (or write code to load it from a file that your notebook can easily find)
#key_path = 'C:/GCP/blackcatdataquery-72aee38227a6.json'

This should not throw an error if key retrieval / assignemnt worked:

In [4]:
#storage_client = storage.Client.from_service_account_json(key_path)

*Nb. The `storage` object was used in the above example, but there are other objects of interest that have polymorphic `Client` members that are used similarly, such as `bigquery`, which is used below.*

If implicit key retrieval is working for you, execute this:

In [5]:
client = bigquery.Client()

if not, execute this:

In [6]:
#client = bigquery.Client.from_service_account_json(key_path)

This client is associated with the default project (which was set or defaulted in the BigQuery UI): 

In [7]:
client.project

'magnetic-math-233501'

A BigQuery project contains datasets. Datasets contain tables. To get at the data in a table we need to create a reference that covers this hierarchy; in the `bigquery` library this looks like `project.dataset.table`.  

(Nb. Queries can be performed on prjects and datasets, but most queries are performed on tables.)

To explore the public datasets we will start by reassgining our `client` variable using optional `project` parameter (set to `bigquery-public-data`):

In [8]:
#project = 'bigquery-public-data'
client = bigquery.Client(project = 'bigquery-public-data')
print(client.project)

bigquery-public-data


Here is how to get a list of the datasets in the current project:

In [9]:
datasets = list(client.list_datasets())
print(datasets)

[<google.cloud.bigquery.dataset.DatasetListItem object at 0x7f6eb070ec18>, <google.cloud.bigquery.dataset.DatasetListItem object at 0x7f6eb070ecf8>, <google.cloud.bigquery.dataset.DatasetListItem object at 0x7f6eb070ed30>, <google.cloud.bigquery.dataset.DatasetListItem object at 0x7f6eb072fe48>, <google.cloud.bigquery.dataset.DatasetListItem object at 0x7f6eb072f208>, <google.cloud.bigquery.dataset.DatasetListItem object at 0x7f6eb072fba8>, <google.cloud.bigquery.dataset.DatasetListItem object at 0x7f6eb07475f8>, <google.cloud.bigquery.dataset.DatasetListItem object at 0x7f6eb07479b0>, <google.cloud.bigquery.dataset.DatasetListItem object at 0x7f6eb0747c88>, <google.cloud.bigquery.dataset.DatasetListItem object at 0x7f6eb0747470>, <google.cloud.bigquery.dataset.DatasetListItem object at 0x7f6eb0747b70>, <google.cloud.bigquery.dataset.DatasetListItem object at 0x7f6eb0747550>, <google.cloud.bigquery.dataset.DatasetListItem object at 0x7f6eb07475c0>, <google.cloud.bigquery.dataset.Datase

That wasn't helpful. We need to go deeper into the object structure to get at something meaningful. Actually, the `dataset_id` member contains the name attribute of a `dataset` object; write some code to print that name for each member of the list that was created above:

In [10]:
for dataset in datasets:
    print(dataset.dataset_id)

austin_311
austin_bikeshare
austin_crime
austin_incidents
austin_waste
baseball
bitcoin_blockchain
bls
census_bureau_construction
census_bureau_international
census_bureau_usa
census_fips_codes
chicago_crime
chicago_taxi_trips
cloud_storage_geo_index
cms_codes
cms_medicare
crypto_bitcoin
crypto_bitcoin_cash
crypto_dash
crypto_dogecoin
crypto_ethereum
crypto_ethereum_classic
crypto_litecoin
crypto_zcash
eclipse_megamovie
epa_historical_air_quality
ethereum_blockchain
fda_drug
fda_food
fec
genomics_cannabis
genomics_rice
geo_us_boundaries
geolite2
ghcn_d
ghcn_m
github_repos
google_analytics_sample
google_political_ads
hacker_news
human_genome_variants
human_variant_annotation
irs_990
libraries_io
london_bicycles
london_crime
london_fire_brigade
medicare
ml_datasets
moon_phases
nasa_wildfire
ncaa_basketball
new_york
new_york_311
new_york_citibike
new_york_mv_collisions
new_york_taxi_trips
new_york_trees
nhtsa_traffic_fatalities
nlm_rxnorm
noaa_goes16
noaa_goes17
noaa_gsod
noaa_hurricanes


The google API objects in the `bigquery` library have their own overloads of the format() function that make them easier to read. Below is a function that exploits the `format` method of `project` and `dataset_id`, providing an easy way to list datasets:

In [11]:
# function for listing datasets in a project:
def printDatasetList(client):
    project = client.project    #: only one project can be associated with a client instance
    datasets = list(client.list_datasets())
    if datasets:
        print('Datasets in project {}:'.format(project))
        for dataset in datasets:  
            print('\t{}'.format(dataset.dataset_id))
        found = True
    else:
        print('{} project does not contain any datasets.'.format(project))
        found = False
    return found

In [12]:
# list datasets in the default project:
flag = printDatasetList(client)  #: assigning to `flag` suppresses printing the return value (normally `True`)

Datasets in project bigquery-public-data:
	austin_311
	austin_bikeshare
	austin_crime
	austin_incidents
	austin_waste
	baseball
	bitcoin_blockchain
	bls
	census_bureau_construction
	census_bureau_international
	census_bureau_usa
	census_fips_codes
	chicago_crime
	chicago_taxi_trips
	cloud_storage_geo_index
	cms_codes
	cms_medicare
	crypto_bitcoin
	crypto_bitcoin_cash
	crypto_dash
	crypto_dogecoin
	crypto_ethereum
	crypto_ethereum_classic
	crypto_litecoin
	crypto_zcash
	eclipse_megamovie
	epa_historical_air_quality
	ethereum_blockchain
	fda_drug
	fda_food
	fec
	genomics_cannabis
	genomics_rice
	geo_us_boundaries
	geolite2
	ghcn_d
	ghcn_m
	github_repos
	google_analytics_sample
	google_political_ads
	hacker_news
	human_genome_variants
	human_variant_annotation
	irs_990
	libraries_io
	london_bicycles
	london_crime
	london_fire_brigade
	medicare
	ml_datasets
	moon_phases
	nasa_wildfire
	ncaa_basketball
	new_york
	new_york_311
	new_york_citibike
	new_york_mv_collisions
	new_york_taxi_trips
	

This list should correspond to what is shown here https://bigquery.cloud.google.com/publicdatasets under the **bigquery-public-data** item.

Here is how to create a dataset reference object by assigning a project and a dataset name:

In [13]:
dataset_id = 'samples'
dataset_ref = client.dataset(dataset_id)

If our current project was something other than `bigquery-public-data`, we could still create this reference by specifying the project that contains the dataset:

In [14]:
dataset_id = 'samples'
dataset_ref = client.dataset(dataset_id, project = 'bigquery-public-data')

How can we get the path of the dataset?

In [15]:
dataset_ref.path

'/projects/bigquery-public-data/datasets/samples'

Explore more of this object's members:

*(HINT: Jupyter Notebooks does not support code completion, but Spyder and other Python IDEs do. If you copy all the above code to a Python file within the IDE, you can type `dataset_ref.` in a new line, then hit the [Tab] key to see the available members for the object.)*

Here is a function for listing the tables in a dataset: 

In [16]:
# function for listing tables in a dataset:
def printTableList(client, dataset_id):
    project = client.project
    dataset_ref = client.dataset(dataset_id, project = project)    
    tables = list(client.list_tables(dataset_ref))
    if tables:
        print('Tables in dataset {}:'.format(dataset_id))
        for table in tables: 
            print('\t{}'.format(table.table_id))
        found = True
    else:
        print('{} dataset does not contain any tables.'.format(dataset_id))
        found = False
    return found

Use this function to list the tables in the current dataset:

In [17]:
printTableList(client, dataset_id)

Tables in dataset samples:
	github_nested
	github_timeline
	gsod
	model
	natality
	shakespeare
	trigrams
	wikipedia


True

To create a reference to a table within the dataset, we use the `table_id` attribute:

In [18]:
table_id = 'shakespeare'
table_ref = dataset_ref.table(table_id)

Check the name of the table that `table_ref` now points to:

In [19]:
table_ref.table_id

'shakespeare'

To access the data in the table itself, we use the `get_table()` method:

In [20]:
table = client.get_table(table_ref)  # API Request

NOTE: The contents of the table are not actually in our memory after this call! We are working with a Big Data platform, now, and we could easily end up pulling GBs or TBs of data by accident. 

To minimise data bandwidth, memory consumption, and processing time, Big Data platforms employ ***lazy evaluation***. This means that no computation or data transfer actually takes place until we *realise* (use) the data. Even if we execute subsequent code that performs calculations on the data, no data flow or computation actually occurs until we request output (e.g. by executing a print to stdout or writing to a file). 

What kind of object is returned by `client.get_table`?

In [21]:
#?
type(table)

google.cloud.bigquery.table.Table

How can we view the design of the table (cOlumn names and types? The name of the boject attribute we need is the same term we learned in the module on databases:

In [22]:
table.schema

[SchemaField('word', 'STRING', 'REQUIRED', 'A single unique word (where whitespace is the delimiter) extracted from a corpus.', ()),
 SchemaField('word_count', 'INTEGER', 'REQUIRED', 'The number of times this word appears in this corpus.', ()),
 SchemaField('corpus', 'STRING', 'REQUIRED', 'The work from which this word was extracted.', ()),
 SchemaField('corpus_date', 'INTEGER', 'REQUIRED', 'The year in which this corpus was published.', ())]

Again, this is messy. If we wanted to refer to the column names and types in code, we might use something like this (which we could then parse into a dict):

In [23]:
result = ["{0} {1}".format(schema.name,schema.field_type) for schema in table.schema]
print(result)

['word STRING', 'word_count INTEGER', 'corpus STRING', 'corpus_date INTEGER']


But if we just want to print them, here is another neat function for that:

In [24]:
# function to print a table schema:
def printTableSchema(aTable):
    schemas = list(aTable.schema)
    if schemas:
        print('Table schema for {}:'.format(aTable.table_id))
        for aSchema in schemas:
            print('\t{0} {1}'.format(aSchema.name, aSchema.field_type))
        found = True
    else:
        found = False
    return found

Use this function to print the table schema:

In [25]:
printTableSchema(table)

Table schema for shakespeare:
	word STRING
	word_count INTEGER
	corpus STRING
	corpus_date INTEGER


True

Now that we know what the columns are, we can write queries. Actually, we construct a query job by assigning an SQL statement to a method of the `client` object:

In [26]:
sql = "SELECT COUNT(1) FROM `bigquery-public-data.samples.shakespeare`"
query_job = client.query(sql)

Forbidden: 403 POST https://www.googleapis.com/bigquery/v2/projects/bigquery-public-data/jobs: Access Denied: Project bigquery-public-data: The user maiiam-service@magnetic-math-233501.iam.gserviceaccount.com does not have bigquery.jobs.create permission in project bigquery-public-data.

Why does this throw an error?

ANSWER: 

My account does not have permission to create data for `big-query-public-data`.
_____________________________________________________

So, what can we do?

In [27]:
# Point client to own project name:

project_name = 'magnetic-math-233501'
client = bigquery.Client(project = project_name)

# Execute query

sql = "SELECT COUNT(1) FROM `bigquery-public-data.samples.shakespeare`"
query_job = client.query(sql)

If that worked, show what query_job is:

In [28]:
type(query_job)

google.cloud.bigquery.job.QueryJob

Once again, due to lazy execution, no actual execution occurs until we request output:

In [29]:
for row in query_job:  # API request - fetches results
    print(row)

Row((164656,), {'f0_': 0})


And, again, we need to manipulate this to make it neat. Each member of the rowset is a list and we only want to extract the value, which is in the first member:

In [30]:
print(row[0])

164656


So, we now know that this table has 164,656 rows. (We would not want to print it!)

A better coding practice is to write SQL statements that assign names (aliases) to derived values, so we don't forget what the resulting rowset contains. Rewite the above SQL statement so that the value returned is aliased a "num_rows", and assign the QueryJob as above:

In [31]:
sql = "SELECT COUNT(1) AS num_rows FROM `bigquery-public-data.samples.shakespeare`"
query_job = client.query(sql)

Now we could use Python's `assert` statement to build a test into the first code block that operates on the rowset:

In [32]:
for row in query_job:  # API request - fetches results
    # Row values can be accessed by field name or index:
    assert row[0] == row.num_rows == row['num_rows']  #: for debugging bad sql
    print(row.num_rows)

164656


The above code checks that the name attribute of the value in `row[0]` is what we expected (i.e. "num_rows"). Also, it shows that we can refer to a field in a row by its object member `num_rows` or by using the same notation we use for Python dictionaries, `['num_rows']`. 

Write, execute, and print the results of a query that fetches 10 rows from the table, each containing the "word", "word_count", and "corpus" fields: 

In [33]:
sql = """
      SELECT word, word_count, corpus 
      FROM `bigquery-public-data.samples.shakespeare`
      LIMIT 10;
      """

query_job = client.query(sql)

In [34]:
for row in query_job:
    assert row[0] == row.word == row['word']
    print(row.word)

LVII
augurs
dimm'd
plagues
treason
surmise
heed
Unthrifty
quality
wherever


(NOTE: Using `assert` religiously is good practice and will make debugging easier, but is probably overkill for non-production code.)

Whenever you catch yourself writing a swag of code to do something that seems rudimentary or low-level, there is a very good chance that you don't need to. A much easier way to handle the above requirement is to use the `to_dataframe` method of the QueryJob object:

In [35]:
df = query_job.to_dataframe()
print(df)    

        word  word_count   corpus
0       LVII           1  sonnets
1     augurs           1  sonnets
2     dimm'd           1  sonnets
3    plagues           1  sonnets
4    treason           1  sonnets
5    surmise           1  sonnets
6       heed           1  sonnets
7  Unthrifty           1  sonnets
8    quality           1  sonnets
9   wherever           1  sonnets


Although the above doesn't use `assert` (which you might still want to include in some test code), you will be able to tell at a glance if something is wrong with the contents of the DataFrame.

#### Final Notes

1. Here is a readable way to code long SQL statements:

In [ ]:
sql = """
    SELECT word, word_count, corpus 
    FROM `bigquery-public-data.samples.shakespeare` 
    LIMIT 10
    """

2. If you had an application that needed to modify the tables or datasets in the `bigquery-public-data` is project, you could copy them to our own project, where you would have the permissions to do as you please with the data (subject to Google's terms of use).

3. We aren't limited to the datasets that are already in BigQuery. We can upload tables from our computer, and we can pull data in from other online souces. We will cover these tasks in another module.

#### Next Steps

If you wish to pick up a few more skills you can go to https://cloud.google.com/bigquery/create-simple-app-api. (Note that we have already been through the preliminaries, so you can start at "Download the sample code".)

Alternatively, you can take a deeper dive into the API here: https://googlecloudplatform.github.io/google-cloud-python/latest/bigquery/usage.html.

## EDA

In [38]:
# import data manip
import pandas as pd

# import data viz
import matplotlib as plt
import seaborn as sns

# import NLP
import nltk 

In [39]:
# download stopwords

nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /home/maianh/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [40]:
from nltk.corpus import stopwords

In [36]:
# Find corpus names in shakespeare

sql = """
    SELECT corpus
    FROM `bigquery-public-data.samples.shakespeare` 
    GROUP BY corpus
    ORDER BY corpus
    LIMIT 10
    """

query_df = client.query(sql).to_dataframe()

query_df

corpus
0          1kinghenryiv
1          1kinghenryvi
2          2kinghenryiv
3          2kinghenryvi
4          3kinghenryvi
5  allswellthatendswell
6    antonyandcleopatra
7           asyoulikeit
8        comedyoferrors
9            coriolanus

In [110]:
# Count words in each corpus

sql = """
    SELECT corpus, SUM(word_count) as num_words
    FROM `bigquery-public-data.samples.shakespeare` 
    GROUP BY corpus
    ORDER BY num_words DESC
    """

num_words_df = client.query(sql).to_dataframe()

num_words_df.head()

corpus  num_words
0          hamlet      32446
1  kingrichardiii      31868
2      coriolanus      29535
3       cymbeline      29231
4    2kinghenryiv      28241

In [37]:
# Count distinct words in each corpus

sql = """
    SELECT corpus, COUNT(DISTINCT word) as distinct_words
    FROM `bigquery-public-data.samples.shakespeare` 
    GROUP BY corpus
    ORDER BY distinct_words DESC
    """

distinct_words_df = client.query(sql).to_dataframe()

distinct_words_df.head()

corpus  distinct_words
0              hamlet            5318
1          kinghenryv            5104
2           cymbeline            4875
3  troilusandcressida            4795
4            kinglear            4784

In [53]:
# Find most common words in each corpus in shakespeare

sql = """
    SELECT corpus, word, SUM(word_count) as total_count
    FROM `bigquery-public-data.samples.shakespeare` 
    GROUP BY corpus, word
    ORDER BY total_count DESC
    LIMIT 10
    """

query_df = client.query(sql).to_dataframe()

In [54]:
query_df

corpus word  total_count
0               hamlet  the          995
1           coriolanus  the          942
2           kinghenryv  the          937
3         2kinghenryiv  the          894
4       kingrichardiii  the          848
5         2kinghenryvi  the          841
6              othello    I          828
7           kinghenryv  and          809
8             kinglear  the          786
9  merrywivesofwindsor    I          782

Unsurprisingly, the top words are all __stop words__, and so, not very meaningful.

In [64]:
# import stopwords from NLTK
eng_stopwords = [word.strip() for word in stopwords.words('english')]

eng_stopwords[1:10]

['me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're"]

In [67]:
eng_title = [word.strip().title() for word in stopwords.words('english')]
eng_upper = [word.strip().upper() for word in stopwords.words('english')]

eng_stopwords = eng_stopwords + eng_title + eng_upper

In [117]:
# Create sql query with stopwords filter

sql = """
    SELECT corpus, LOWER(word), SUM(word_count) as total_count
    FROM `bigquery-public-data.samples.shakespeare` 
    WHERE LOWER(word) NOT IN ({})
    GROUP BY corpus, LOWER(word)
    ORDER BY total_count DESC, corpus ASC
    """

print(sql.format(str(eng_stopwords)[1:-1]))


    SELECT corpus, LOWER(word), SUM(word_count) as total_count
    FROM `bigquery-public-data.samples.shakespeare` 
    WHERE LOWER(word) NOT IN ('i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why

In [119]:
# Find most common words in each corpus, omitting stop words

word_freq_df = client.query(sql.format(str(eng_stopwords)[1:-1])).to_dataframe()

word_freq_df.head(15)

corpus       f0_  total_count
0                hamlet    hamlet          484
1          twelfthnight       sir          453
2    antonyandcleopatra    antony          419
3          3kinghenryvi      king          396
4   merrywivesofwindsor  mistress          396
5          juliuscaesar    brutus          371
6               othello      iago          361
7   merrywivesofwindsor      page          358
8               othello   othello          339
9        romeoandjuliet     romeo          322
10        timonofathens     timon          322
11           kinghenryv      king          315
12               hamlet      lord          312
13             kinglear      king          311
14              macbeth   macbeth          311

In [108]:
top_word = word_freq_df.groupby(['corpus']).head(1)

top_word

corpus         f0_  total_count
0                    hamlet      hamlet          484
1              twelfthnight         sir          453
2        antonyandcleopatra      antony          419
3              3kinghenryvi        king          396
4       merrywivesofwindsor    mistress          396
5              juliuscaesar      brutus          371
6                   othello        iago          361
9            romeoandjuliet       romeo          322
10            timonofathens       timon          322
11               kinghenryv        king          315
13                 kinglear        king          311
14                  macbeth     macbeth          311
16           kingrichardiii        king          300
19              asyoulikeit    rosalind          274
21            kingrichardii        king          271
22                  sonnets         thy          266
23       troilusandcressida     troilus          264
25       measureforemeasure        duke          260
28             1kinghenryiv       henry          255
32          titusandronicus  andronicus          248
35               coriolanus  coriolanus          242
36                 kingjohn        king          242
37             2kinghenryvi        king          241
40           comedyoferrors    syracuse          232
45             2kinghenryiv        lord          223
49            kinghenryviii        king          221
53     twogentlemenofverona     proteus          218
55     allswellthatendswell        lord          212
56     periclesprinceoftyre    pericles          212
60         tamingoftheshrew   petruchio          208
69      muchadoaboutnothing    benedick          199
75         loveslabourslost       biron          194
80                  tempest        thou          184
91             1kinghenryvi        thou          174
107               cymbeline        thou          165
127        merchantofvenice      portia          148
131             winterstale     leontes          144
236   midsummersnightsdream        thou          116
245           rapeoflucrece         thy          114
638          venusandadonis        thou           68
1448                various        love           42
8759        loverscomplaint       would           10

Some works have _thou_ or _thy_ as one of the most common words. Again, this is not suprising: it's Shakespeare! These words are not part of the stop words list but probably should be included for this dataset.

In [120]:
# Add old pronouns to stopwords and execute new query

archaic_pronouns = ['thou', 'thy', 'thine']

eng_stopwords = eng_stopwords + archaic_pronouns

sql = """
    SELECT corpus, LOWER(word), SUM(word_count) as total_count
    FROM `bigquery-public-data.samples.shakespeare` 
    WHERE LOWER(word) NOT IN ({})
    GROUP BY corpus, LOWER(word)
    ORDER BY total_count DESC, corpus ASC
    """

word_freq_df = client.query(sql.format(str(eng_stopwords)[1:-1])).to_dataframe()

top_word = word_freq_df.groupby(['corpus']).head(1)

top_word

corpus         f0_  total_count
0                    hamlet      hamlet          484
1              twelfthnight         sir          453
2        antonyandcleopatra      antony          419
3              3kinghenryvi        king          396
4       merrywivesofwindsor    mistress          396
5              juliuscaesar      brutus          371
6                   othello        iago          361
9            romeoandjuliet       romeo          322
10            timonofathens       timon          322
11               kinghenryv        king          315
13                 kinglear        king          311
14                  macbeth     macbeth          311
16           kingrichardiii        king          300
18              asyoulikeit    rosalind          274
20            kingrichardii        king          271
21       troilusandcressida     troilus          264
23       measureforemeasure        duke          260
26             1kinghenryiv       henry          255
29          titusandronicus  andronicus          248
32               coriolanus  coriolanus          242
33                 kingjohn        king          242
34             2kinghenryvi        king          241
36           comedyoferrors    syracuse          232
41             2kinghenryiv        lord          223
45            kinghenryviii        king          221
47     twogentlemenofverona     proteus          218
49     allswellthatendswell        lord          212
50     periclesprinceoftyre    pericles          212
52         tamingoftheshrew   petruchio          208
58      muchadoaboutnothing    benedick          199
63         loveslabourslost       biron          194
78             1kinghenryvi      talbot          167
81                  sonnets        love          167
87                cymbeline      imogen          162
102        merchantofvenice      portia          148
105             winterstale     leontes          144
108                 tempest    prospero          142
229   midsummersnightsdream        love          109
554           rapeoflucrece        doth           69
572          venusandadonis           '           68
1370                various        love           42
8655        loverscomplaint       would           10

Many top words seem to be shared across Shakespeare's works. For example, _king_ is the top word for Macbeth and King Lear. Other top words are character names, most often the protagonist of the work.  

In [123]:
top_word.sort_values(['f0_', 'total_count'])

corpus         f0_  total_count
572          venusandadonis           '           68
29          titusandronicus  andronicus          248
2        antonyandcleopatra      antony          419
58      muchadoaboutnothing    benedick          199
63         loveslabourslost       biron          194
5              juliuscaesar      brutus          371
32               coriolanus  coriolanus          242
554           rapeoflucrece        doth           69
23       measureforemeasure        duke          260
0                    hamlet      hamlet          484
26             1kinghenryiv       henry          255
6                   othello        iago          361
87                cymbeline      imogen          162
45            kinghenryviii        king          221
34             2kinghenryvi        king          241
33                 kingjohn        king          242
20            kingrichardii        king          271
16           kingrichardiii        king          300
13                 kinglear        king          311
11               kinghenryv        king          315
3              3kinghenryvi        king          396
105             winterstale     leontes          144
49     allswellthatendswell        lord          212
41             2kinghenryiv        lord          223
1370                various        love           42
229   midsummersnightsdream        love          109
81                  sonnets        love          167
14                  macbeth     macbeth          311
4       merrywivesofwindsor    mistress          396
50     periclesprinceoftyre    pericles          212
52         tamingoftheshrew   petruchio          208
102        merchantofvenice      portia          148
108                 tempest    prospero          142
47     twogentlemenofverona     proteus          218
9            romeoandjuliet       romeo          322
18              asyoulikeit    rosalind          274
1              twelfthnight         sir          453
36           comedyoferrors    syracuse          232
78             1kinghenryvi      talbot          167
10            timonofathens       timon          322
21       troilusandcressida     troilus          264
8655        loverscomplaint       would           10

In [115]:
# Join top words with num words

shakes_words = pd.merge(num_words_df, top_word, on='corpus')
shakes_words['freq'] = shakes_words['total_count'] / shakes_words['num_words'] * 100

shakes_words.sort_values('freq')

corpus  num_words         f0_  total_count      freq
41        loverscomplaint       2586       would           10  0.386698
15            winterstale      26181     leontes          144  0.550017
3               cymbeline      29231        thou          165  0.564469
24       merchantofvenice      22448      portia          148  0.659301
36  midsummersnightsdream      17348        thou          116  0.668665
39         venusandadonis      10035        thou           68  0.677628
20           1kinghenryvi      23272        thou          174  0.747680
38          rapeoflucrece      15221         thy          114  0.748965
4            2kinghenryiv      28241        lord          223  0.789632
2              coriolanus      29535  coriolanus          242  0.819367
21       loveslabourslost      23189       biron          194  0.836604
13          kinghenryviii      26265        king          221  0.841424
16   allswellthatendswell      24622        lord          212  0.861019
23    muchadoaboutnothing      22760    benedick          199  0.874341
10           2kinghenryvi      27165        king          241  0.887171
25       tamingoftheshrew      22358   petruchio          208  0.930316
1          kingrichardiii      31868        king          300  0.941383
8      troilusandcressida      27837     troilus          264  0.948378
11           1kinghenryiv      26401       henry          255  0.965873
35                tempest      17593        thou          184  1.045871
30   periclesprinceoftyre      19846    pericles          212  1.068225
26               kingjohn      21983        king          242  1.100851
6                kinglear      28001        king          311  1.110675
19     measureforemeasure      23303        duke          260  1.115736
17          kingrichardii      24150        king          271  1.122153
7              kinghenryv      27894        king          315  1.129275
27        titusandronicus      21911  andronicus          248  1.131852
32   twogentlemenofverona      18493     proteus          218  1.178824
40                various       3545        love           42  1.184767
22            asyoulikeit      23087    rosalind          274  1.186815
14         romeoandjuliet      26192       romeo          322  1.229383
5                 othello      28076        iago          361  1.285796
37         comedyoferrors      16361    syracuse          232  1.418006
0                  hamlet      32446      hamlet          484  1.491709
34                sonnets      17805         thy          266  1.493962
12           3kinghenryvi      26288        king          396  1.506391
9      antonyandcleopatra      27310      antony          419  1.534237
31          timonofathens      19841       timon          322  1.622902
18    merrywivesofwindsor      24033    mistress          396  1.647734
33                macbeth      18439     macbeth          311  1.686642
29           juliuscaesar      21052      brutus          371  1.762303
28           twelfthnight      21633         sir          453  2.094023

## - END -

>
>


>
>




---



---

> > > > > > > > > © 2019 Data Science Institute of Australia


---



---

